File: tests/Scribe.Core.Tests/DictionaryCsvTests.cs
using Scribe.Core.Models;
using Scribe.Core.PostProcessing;
using Xunit;

namespace Scribe.Core.Tests;

public class DictionaryCsvTests
{
    [Fact]
    public void Export_then_parse_round_trips_entries()
    {
        var entries = new[]
        {
            new DictionaryEntry(1, "azure", "Azure"),
            new DictionaryEntry(2, "cube flow", "Kubeflow", WholeWord: false),
            new DictionaryEntry(3, "kay eight ess", "K8s", Enabled: false),
        };

        var result = DictionaryCsv.Parse(DictionaryCsv.Export(entries));

        Assert.Empty(result.Errors);
        Assert.Equal(3, result.Entries.Count);
        // Imported entries are new rows (Id 0); everything else survives the trip.
        Assert.All(result.Entries, e => Assert.Equal(0, e.Id));
        Assert.Equal(("azure", "Azure", true, true),
            (result.Entries[0].Pattern, result.Entries[0].Replacement, result.Entries[0].WholeWord, result.Entries[0].Enabled));
        Assert.Equal(("cube flow", "Kubeflow", false, true),
            (result.Entries[1].Pattern, result.Entries[1].Replacement, result.Entries[1].WholeWord, result.Entries[1].Enabled));
        Assert.Equal(("kay eight ess", "K8s", true, false),
            (result.Entries[2].Pattern, result.Entries[2].Replacement, result.Entries[2].WholeWord, result.Entries[2].Enabled));
    }

    [Fact]
    public void Quoted_fields_with_commas_and_quotes_round_trip()
    {
        var entries = new[] { new DictionaryEntry(1, "acme, inc", "Acme \"The Best\" Inc.") };

        var result = DictionaryCsv.Parse(DictionaryCsv.Export(entries));

        Assert.Empty(result.Errors);
        var entry = Assert.Single(result.Entries);
        Assert.Equal("acme, inc", entry.Pattern);
        Assert.Equal("Acme \"The Best\" Inc.", entry.Replacement);
    }

    [Fact]
    public void Comments_blank_lines_and_header_are_skipped()
    {
        const string csv =
            """
            # a comment, with a comma
            pattern,replacement,whole_word,enabled

            azure,Azure
            """;

        var result = DictionaryCsv.Parse(csv);

        Assert.Empty(result.Errors);
        var entry = Assert.Single(result.Entries);
        Assert.Equal("azure", entry.Pattern);
        Assert.True(entry.WholeWord);  // optional column defaults
        Assert.True(entry.Enabled);
    }

    [Fact]
    public void The_shipped_template_parses_cleanly()
    {
        var result = DictionaryCsv.Parse(DictionaryCsv.Template);

        Assert.Empty(result.Errors);
        Assert.Equal(3, result.Entries.Count);
        Assert.Contains(result.Entries, e => e.Pattern == "cube flow" && e.Replacement == "Kubeflow");
    }

    [Fact]
    public void Flag_columns_accept_yes_no_and_numeric_forms()
    {
        const string csv =
            """
            one,One,no,0
            two,Two,YES,1
            """;

        var result = DictionaryCsv.Parse(csv);

        Assert.Empty(result.Errors);
        Assert.False(result.Entries[0].WholeWord);
        Assert.False(result.Entries[0].Enabled);
        Assert.True(result.Entries[1].WholeWord);
        Assert.True(result.Entries[1].Enabled);
    }

    [Fact]
    public void Bad_rows_are_reported_with_line_numbers_and_good_rows_still_import()
    {
        const string csv =
            """
            azure,Azure
            just-a-pattern
            ,MissingPattern
            foundry,Foundry,maybe
            rebac,ReBAC
            """;

        var result = DictionaryCsv.Parse(csv);

        Assert.Equal(2, result.Entries.Count); // azure + rebac survive
        Assert.Equal(3, result.Errors.Count);
        Assert.Contains(result.Errors, e => e.StartsWith("Line 2:"));
        Assert.Contains(result.Errors, e => e.StartsWith("Line 3:"));
        Assert.Contains(result.Errors, e => e.StartsWith("Line 4:") && e.Contains("maybe"));
    }

    [Fact]
    public void Empty_or_null_input_yields_nothing()
    {
        Assert.Empty(DictionaryCsv.Parse(null).Entries);
        Assert.Empty(DictionaryCsv.Parse("  \r\n ").Entries);
        Assert.Empty(DictionaryCsv.Parse(string.Empty).Errors);
    }

    [Fact]
    public void Quoted_field_containing_a_line_break_stays_one_record()
    {
        // Spreadsheets can emit embedded line breaks in quoted cells; the reader must not split
        // the record. (The glossary/post-processor flatten the value later; parsing just has to
        // survive it.)
        const string csv = "\"multi\nline\",Value";

        var result = DictionaryCsv.Parse(csv);

        Assert.Empty(result.Errors);
        var entry = Assert.Single(result.Entries);
        Assert.Equal("multi\nline", entry.Pattern);
        Assert.Equal("Value", entry.Replacement);
    }
}

File: macos/Scribe/Tests/ScribeTests/DictionaryCsvTests.swift
import XCTest
@testable import Scribe

final class DictionaryCsvTests: XCTestCase {
    func testExportThenParseRoundTripsEntries() {
        let entries = [
            DictionaryEntry(id: 1, pattern: "azure", replacement: "Azure"),
            DictionaryEntry(id: 2, pattern: "cube flow", replacement: "Kubeflow", wholeWord: false),
            DictionaryEntry(id: 3, pattern: "kay eight ess", replacement: "K8s", enabled: false),
        ]

        let result = DictionaryCsv.parse(DictionaryCsv.export(entries))

        XCTAssertTrue(result.errors.isEmpty)
        XCTAssertEqual(result.entries.count, 3)
        // Imported entries are new rows (id 0); everything else survives the trip.
        XCTAssertTrue(result.entries.allSatisfy { $0.id == 0 })
        XCTAssertEqual(result.entries[0].pattern, "azure")
        XCTAssertEqual(result.entries[0].replacement, "Azure")
        XCTAssertTrue(result.entries[0].wholeWord)
        XCTAssertTrue(result.entries[0].enabled)

        XCTAssertEqual(result.entries[1].pattern, "cube flow")
        XCTAssertEqual(result.entries[1].replacement, "Kubeflow")
        XCTAssertFalse(result.entries[1].wholeWord)
        XCTAssertTrue(result.entries[1].enabled)

        XCTAssertEqual(result.entries[2].pattern, "kay eight ess")
        XCTAssertEqual(result.entries[2].replacement, "K8s")
        XCTAssertTrue(result.entries[2].wholeWord)
        XCTAssertFalse(result.entries[2].enabled)
    }

    func testQuotedFieldsWithCommasAndQuotesRoundTrip() {
        let entries = [DictionaryEntry(id: 1, pattern: "acme, inc", replacement: "Acme \"The Best\" Inc.")]

        let result = DictionaryCsv.parse(DictionaryCsv.export(entries))

        XCTAssertTrue(result.errors.isEmpty)
        XCTAssertEqual(result.entries.count, 1)
        XCTAssertEqual(result.entries[0].pattern, "acme, inc")
        XCTAssertEqual(result.entries[0].replacement, "Acme \"The Best\" Inc.")
    }

    func testCommentsBlankLinesAndHeaderAreSkipped() {
        let csv = """
            # a comment, with a comma
            pattern,replacement,whole_word,enabled

            azure,Azure
            """

        let result = DictionaryCsv.parse(csv)

        XCTAssertTrue(result.errors.isEmpty)
        XCTAssertEqual(result.entries.count, 1)
        XCTAssertEqual(result.entries[0].pattern, "azure")
        XCTAssertTrue(result.entries[0].wholeWord) // optional column defaults
        XCTAssertTrue(result.entries[0].enabled)
    }

    func testTheShippedTemplateParsesCleanly() {
        let result = DictionaryCsv.parse(DictionaryCsv.template)

        XCTAssertTrue(result.errors.isEmpty)
        XCTAssertEqual(result.entries.count, 3)
        XCTAssertTrue(result.entries.contains { $0.pattern == "cube flow" && $0.replacement == "Kubeflow" })
    }

    func testFlagColumnsAcceptYesNoAndNumericForms() {
        let csv = """
            one,One,no,0
            two,Two,YES,1
            """

        let result = DictionaryCsv.parse(csv)

        XCTAssertTrue(result.errors.isEmpty)
        XCTAssertFalse(result.entries[0].wholeWord)
        XCTAssertFalse(result.entries[0].enabled)
        XCTAssertTrue(result.entries[1].wholeWord)
        XCTAssertTrue(result.entries[1].enabled)
    }

    func testBadRowsAreReportedWithLineNumbersAndGoodRowsStillImport() {
        let csv = """
            azure,Azure
            just-a-pattern
            ,MissingPattern
            foundry,Foundry,maybe
            rebac,ReBAC
            """

        let result = DictionaryCsv.parse(csv)

        XCTAssertEqual(result.entries.count, 2) // azure + rebac survive
        XCTAssertEqual(result.errors.count, 3)
        XCTAssertTrue(result.errors.contains { $0.hasPrefix("Line 2:") })
        XCTAssertTrue(result.errors.contains { $0.hasPrefix("Line 3:") })
        XCTAssertTrue(result.errors.contains { $0.hasPrefix("Line 4:") && $0.contains("maybe") })
    }

    func testEmptyOrNilInputYieldsNothing() {
        XCTAssertTrue(DictionaryCsv.parse(nil).entries.isEmpty)
        XCTAssertTrue(DictionaryCsv.parse("  \r\n ").entries.isEmpty)
        XCTAssertTrue(DictionaryCsv.parse("").errors.isEmpty)
    }

    func testQuotedFieldContainingALineBreakStaysOneRecord() {
        // Spreadsheets can emit embedded line breaks in quoted cells; the reader must not split
        // the record.
        let csv = "\"multi\nline\",Value"

        let result = DictionaryCsv.parse(csv)

        XCTAssertTrue(result.errors.isEmpty)
        XCTAssertEqual(result.entries.count, 1)
        XCTAssertEqual(result.entries[0].pattern, "multi\nline")
        XCTAssertEqual(result.entries[0].replacement, "Value")
    }

    func testUnterminatedQuotedFieldIsRejectedWithoutAbsorbingFollowingRows() {
        let csv = "azure,\"Azure\nrebac,ReBAC"

        let result = DictionaryCsv.parse(csv)

        XCTAssertTrue(result.entries.isEmpty)
        XCTAssertTrue(result.errors.contains { $0.contains("closing quote") })
    }
}
